frontend/ui/widget/menu.lua

```diff
@@ -1176,7 +1185,71 @@ function Menu:getPageNumber(item_number)
     if #self.item_table == 0 or item_number == 0 then
         return 1
     end
-    return math.ceil(math.min(item_number, #self.item_table) / self.perpage)
+    if self.items_max_lines then
+        for page, items in ipairs(self.page_items) do
+            if item_number <= items[#items] then
+                return page
+            end
+        end
+        return #self.page_items
+    else
+        return math.ceil(math.min(item_number, #self.item_table) / self.perpage)
+    end
+end
+
+function Menu:setupItemHeights()
+    if #self.item_table == 0 then
+        self.page_items = {{}}
+        return
+    end
+
+    local face = Font:getFace("smallinfofont", self.font_size)
+    local line_height = TextBoxWidget:new{
+        text = "A",
+        face = face,
+    }:getSize().h
+    local infont_size = self.items_mandatory_font_size or (self.font_size - 4)
+    local infont_face = Font:getFace("infont", infont_size)
+    local infont_char_width = TextWidget:new{
+        text = "0",
+        face = infont_face,
+        bold = true,
+    }:getSize().w
```

Comment:
So, your below calculation `infont_char_width * (item.mandatory and #item.mandatory or 0)` works because the font is monospace, right ? So no need to measure/reader each mandatory to get its width.
We have to trust the monospace font to be really monospace :) and be sure there is not "hidden/ignorable" codepoint in the `item.mandatory` string like bidi chars or narrow spaces. Are we sure of that? Is it always trings we build with our numbers, and never strings we get from the book?

Oh, I remember https://github.com/koreader/koreader/issues/11719#issuecomment-2081369184 - so we also need to be sure there won't be any diacritic in the mandatory, coming for example from translators translation of our ASCII english strings.

Reply:
The only purpose is to speed up. I try to avoid building the correct TextWidget for each mandatory, it slows down twice.
I tried giving 10% of the full width to the mandatory, but it is too rough.
So I estimate as above. Better ideas are welcome.

Reply:
I have no better idea about this point and the one below :/ So, ok.